datajoint/base_relation.py

```diff
@@ -341,35 +349,42 @@ def delete(self):
             if restrictions[name]:  # do not restrict by an empty list
                 r.restrict([r.proj() if isinstance(r, RelationalOperand) else r
                             for r in restrictions[name]])
-        # execute
-        do_delete = False  # indicate if there is anything to delete
-        if config['safemode']:  # pragma: no cover
-            print('The contents of the following tables are about to be deleted:')
-
-        for table, relation in list(delete_list.items()):   # need list to force a copy
-            if table.isdigit():
-                delete_list.pop(table)  # remove alias nodes from the delete list
-            else:
-                count = len(relation)
-                if count:
-                    do_delete = True
-                    if config['safemode']:
-                        print(table, '(%d tuples)' % count)
-                else:
-                    delete_list.pop(table)
-        if not do_delete:
-            if config['safemode']:
-                print('Nothing to delete')
+        if safe:
+            print('About to delete:')
+
+        if not already_in_transaction:
+            self.connection.start_transaction()
+        total = 0
+        try:
+            for r in reversed(list(delete_list.values())):
+                count = r.delete_quick(get_count=True)
+                total += count
+                if (verbose or safe) and count:
+                    print('{table}: {count} items'.format(table=r.full_table_name, count=count))
+        except:
+            # Delete failed, perhaps due to insufficient privileges. Cancel transaction.
+            if not already_in_transaction:
+                self.connection.cancel_transaction()
+            raise
         else:
-            if not config['safemode'] or user_choice("Proceed?", default='no') == 'yes':
-                already_in_transaction = self.connection._in_transaction
-                if not already_in_transaction:
-                    self.connection.start_transaction()
-                for r in reversed(list(delete_list.values())):
-                    r.delete_quick()
+            assert not (already_in_transaction and safe)
+            if not total:
+                print('Nothing to delete')
                 if not already_in_transaction:
-                    self.connection.commit_transaction()
-                print('Done')
+                    self.connection.cancel_transaction()
+            else:
+                if already_in_transaction:
+                    if verbose:
+                        print('The delete is pending within the ongoing transaction.')
+                else:
+                    if not safe or user_choice("Proceed?", default='no') == 'yes':
```

Comment:
My one concern with this approach is that it has the potential to lock the table when waiting for user input..
I suppose this is a tradeoff between potentially inconsistent prompting notification and actual delete actions..
Don't think it is a reason to block and I think I do prefer this approach, but thought I'd mention

Reply:
The blocking is a bit more complex than this.  Block would have occurred under the previous approach as well after the consistent read was made.